libs/cardano-data/src/Data/OSet/Strict.hs

```diff
@@ -0,0 +1,152 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Data.OSet.Strict (
+  OSet (Empty, (:<|:), (:|>:)),
+  size,
+  empty,
+  singleton,
+  lookup,
+  (!?),
+  fromStrictSeq,
+  toStrictSeq,
+  (|>),
+  (<|),
+  (|><),
+  (><|),
+  member,
+  uncons,
+  unsnoc,
+)
+where
+
+import Cardano.Binary (FromCBOR (..), ToCBOR (..))
+import Codec.Serialise (Serialise)
+import Control.DeepSeq (NFData)
+import Data.Aeson (FromJSON (..), ToJSON (..))
+import Data.Foldable (foldl')
+import Data.Sequence qualified as Seq
+import Data.Sequence.Strict qualified as SSeq
+import NoThunks.Class (NoThunks (..))
+import Prelude hiding (lookup, null)
+
+-- | A general-purpose finite, ordered, set that is strict in its values.
+--
+-- The strictness is enforced by the underlying `StrictSeq` from base, and the
+-- uniqueness of the values is enforced in this module, by the constructing
+-- functions rather than at the type-level.
+newtype OSet a = OSet {unOSet :: SSeq.StrictSeq a}
+  deriving stock (Eq, Ord, Show)
+  deriving newtype
+    ( Serialise
+    , NFData
+    , ToCBOR
+    , FromCBOR
```

Comment:
These should not be derived. We need decoders that fail when duplicates are encoded
```suggestion
```

Reply:
Noted! :raised_hands: Thank you! :bow: